sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/MacPool.java

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.implementation;
+
+import javax.crypto.Mac;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+class MacPool {
+    final Mac macInstance;
+    final ConcurrentLinkedQueue<Mac> pool;
```

Comment:
Using MacPool is a very nice change.

however now mac.clone() overhead is replaced by a synchrnozation lock. most likely better perf. but do we know how much perf gain?

You can do micro-benchmarking using jmh (just for the authorization part) I have the jmh project template here:
https://github.com/moderakh/azure-cosmosdb-benchmark/tree/master/jmh

Reply:
Won't this benchmark needs package to be published?

Reply:
no, you don't need to publish the package. There are two options:

first option)
 on the java SDK repo install the package in your local maven repo using following command:
```mvn -e  -Dgpg.skip -Dmaven.javadoc.skip=true -Dspotbugs.skip=true  -Dcheckstyle.skip=true  -Drevapi.skip=true  -DskipTests -pl ,com.azure:azure-cosmos  -am clean install```
and in the jmh project just update the pom to use the latest version installed in your local maven repo.
example for this option in jmh project:
https://github.com/moderakh/azure-cosmosdb-benchmark/blob/master/jmh/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdTokenStreamJMH.java

second option)
if the area to be tested is small (just copy that code (e.g, MacPool) to jmh project itself and micro benchmark just the copied code.
example for this option:
copied code:
https://github.com/moderakh/azure-cosmosdb-benchmark/blob/master/jmh/src/main/java/moderakh/ConsistencyLevel.java
micro-benchmark test:
https://github.com/moderakh/azure-cosmosdb-benchmark/blob/master/jmh/src/main/java/moderakh/ExpandableStringEnumJMH.java

Reply:
Based on offline discussion resolving this comment.